Error when using vcc  and cpp backend. The emitted cpp errors converting NI16* to const wchar_t*.

### Description

vcctest.nim:
```nim
import std/os
echo getEnv("Path")
```

nim c --cc:vcc --backend:cpp vcctest.nim

### Nim Version

Nim Compiler Version 1.9.1 [Windows: amd64]
Compiled at 2022-12-16
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 7fa983f5e6b5f52a20e477e2443c8fdede244290
active boot switches: -d:release

### Current Output

```text
PS C:\nimdev\_experiments> nim r -f vcctest
Hint: used config file 'C:\Nim\config\nim.cfg' [Conf]
Hint: used config file 'C:\Nim\config\config.nims' [Conf]
Hint: used config file 'vcctest.nim.cfg' [Conf]
..................................................................................................................
CC: ../../Nim/lib/system/exceptions.nim
CC: ../../Nim/lib/std/sysatomics.nim
CC: ../../Nim/lib/std/private/digitsutils.nim
CC: ../../Nim/lib/system/dollars.nim
CC: ../../Nim/lib/std/exitprocs.nim
CC: ../../Nim/lib/std/syncio.nim
CC: ../../Nim/lib/system.nim
CC: ../../Nim/lib/pure/dynlib.nim
CC: ../../Nim/lib/windows/winlean.nim
CC: ../../Nim/lib/pure/times.nim
CC: ../../Nim/lib/std/private/win_setenv.nim
CC: ../../Nim/lib/std/envvars.nim
CC: ../../Nim/lib/std/cmdline.nim
CC: vcctest.nim
@m..@s..@sNim@slib@sstd@ssysatomics.nim.cpp
@m..@s..@sNim@slib@ssystem@sexceptions.nim.cpp
@m..@s..@sNim@slib@sstd@swidestrs.nim.cpp
@m..@s..@sNim@slib@sstd@sprivate@sdigitsutils.nim.cpp
@m..@s..@sNim@slib@ssystem.nim.cpp
@m..@s..@sNim@slib@sstd@sexitprocs.nim.cpp
@m..@s..@sNim@slib@ssystem@sdollars.nim.cpp
@m..@s..@sNim@slib@spure@sdynlib.nim.cpp
@mvcctest.nim.cpp
@m..@s..@sNim@slib@sstd@senvvars.nim.cpp
@m..@s..@sNim@slib@spure@stimes.nim.cpp
@m..@s..@sNim@slib@sstd@ssyncio.nim.cpp
C:\Users\geekr\nimcache\vcctest_d\@m..@s..@sNim@slib@sstd@senvvars.nim.cpp(226): error C2664: 'wchar_t *_wgetenv(const wchar_t *)': cannot convert argument 1 from 'NI16 *' to 'const wchar_t *'
C:\Users\geekr\nimcache\vcctest_d\@m..@s..@sNim@slib@sstd@senvvars.nim.cpp(226): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or parenthesized function-style cast       
C:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\ucrt\corecrt_wstdlib.h(418): note: see declaration of '_wgetenv'
C:\Users\geekr\nimcache\vcctest_d\@m..@s..@sNim@slib@sstd@senvvars.nim.cpp(226): note: while trying to match the argument list '(NI16 *)'
@m..@s..@sNim@slib@swindows@swinlean.nim.cpp
@m..@s..@sNim@slib@sstd@sprivate@swin_setenv.nim.cpp
@m..@s..@sNim@slib@sstd@scmdline.nim.cpp
Error: execution of an external compiler program 'vccexe.exe /c --platform:amd64  /nologo /nologo /EHsc -DWIN32_LEAN_AND_MEAN   /IC:\Nim\lib /IC:\nimdev\_experiments /nologo /FoC:\Users\geekr\nimcache\vcctest_d\@m..@s..@sNim@slib@sstd@senvvars.nim.cpp.obj C:\Users\geekr\nimcache\vcctest_d\@m..@s..@sNim@slib@sstd@senvvars.nim.cpp' failed with exit code: 2
```


### Expected Output

```text
Should compile.
```


### Possible Solution

_No response_

### Additional Information

This compiles with gcc, but a project may require vcc as a dependency.

--- a/lib/std/envvars.nim
+++ b/lib/std/envvars.nim
@@ -63,9 +63,13 @@ when not defined(nimscript):
       import winlean
       when defined(nimPreviewSlimSystem):
         import std/widestrs
-      proc c_wgetenv(varname: WideCString): WideCString {.importc: "_wgetenv",
+
+      type wchar_t {.importc: "wchar_t", header: "<stdlib.h>".} = int16
+      proc c_wgetenv(varname: ptr wchar_t): ptr wchar_t {.importc: "_wgetenv",
           header: "<stdlib.h>".}
-      proc getEnvImpl(env: cstring): WideCString = c_wgetenv(env.newWideCString)
+      proc getEnvImpl(env: cstring): WideCString =
+        let r: WideCString = env.newWideCString
+        cast[WideCString](c_wgetenv(cast[ptr wchar_t](r)))
     else:
       proc c_getenv(env: cstring): cstring {.
         importc: "getenv", header: "<stdlib.h>".}

--- a/lib/std/private/win_setenv.nim
+++ b/lib/std/private/win_setenv.nim
@@ -33,25 +33,25 @@ else:
     # same as winlean.setEnvironmentVariableA
 
   proc c_getenv(varname: cstring): cstring {.importc: "getenv", header: "<stdlib.h>".}
-  proc c_wputenv(envstring: WideCString): cint {.importc: "_wputenv", header: "<stdlib.h>".}
-  proc c_wgetenv(varname: WideCString): WideCString {.importc: "_wgetenv", header: "<stdlib.h>".}
+  proc c_wputenv(envstring: ptr wchar_t): cint {.importc: "_wputenv", header: "<stdlib.h>".}
+  proc c_wgetenv(varname: ptr wchar_t): ptr wchar_t {.importc: "_wgetenv", header: "<stdlib.h>".}
 
   var errno {.importc, header: "<errno.h>".}: cint
   var genviron {.importc: "_environ".}: ptr ptr char
     # xxx `ptr UncheckedArray[WideCString]` did not work
 
-  proc wcstombs(wcstr: ptr char, mbstr: WideCString, count: csize_t): csize_t {.importc, header: "<stdlib.h>".}
+  proc wcstombs(wcstr: ptr char, mbstr: ptr wchar_t, count: csize_t): csize_t {.importc, header: "<stdlib.h>".}
     # xxx cint vs errno_t?
 
   proc setEnvImpl*(name: string, value: string, overwrite: cint): cint =
     const EINVAL = cint(22)
-    let wideName = newWideCString(name)
-    if overwrite == 0 and c_wgetenv(wideName) != nil:
+    let wideName: WideCString = newWideCString(name)
+    if overwrite == 0 and c_wgetenv(cast[ptr wchar_t](wideName)) != nil:
       return 0
 
     if value != "":
-      let envstring = name & "=" & value
-      let e = c_wputenv(newWideCString(envstring))
+      let envstring: WideCString = newWideCString(name & "=" & value)
+      let e = c_wputenv(cast[ptr wchar_t](envstring))
       if e != 0:
         errno = EINVAL
         return -1
@@ -62,19 +62,19 @@ else:
     SetEnvironmentVariableA doesn't update `_environ`,
     so we have to do these terrible things.
     ]#
-    let envstring = name & "=  "
-    if c_wputenv(newWideCString(envstring)) != 0:
+    let envstring: WideCString = newWideCString(name & "=  ")
+    if c_wputenv(cast[ptr wchar_t](envstring)) != 0:
       errno = EINVAL
       return -1
     # Here lies the documentation we blatently ignore to make this work.
-    var s = c_wgetenv(wideName)
+    var s = cast[WideCString](c_wgetenv(cast[ptr wchar_t](wideName)))
     s[0] = Utf16Char('\0')
     #[
     This would result in a double null termination, which normally signifies the
     end of the environment variable list, so we stick a completely empty
     environment variable into the list instead.
     ]#
-    s = c_wgetenv(wideName)
+    s = cast[WideCString](c_wgetenv(cast[ptr wchar_t](wideName)))
     s[1] = Utf16Char('=')
     #[
     If genviron is null, the MBCS environment has not been initialized
@@ -88,12 +88,12 @@ else:
       # in the current codepage. Skip updating MBCS environment in this case.
       # For some reason, second `wcstombs` can find non-convertible characters
       # that the first `wcstombs` cannot.
-      let requiredSizeS = wcstombs(nil, wideName, 0)
+      let requiredSizeS = wcstombs(nil, cast[ptr wchar_t](wideName), 0)
       if requiredSizeS != high(csize_t):
         let requiredSize = requiredSizeS.int
         var buf = newSeq[char](requiredSize + 1)
         let buf2 = buf[0].addr
-        if wcstombs(buf2, wideName, csize_t(requiredSize + 1)) != high(csize_t):
+        if wcstombs(buf2, cast[ptr wchar_t](wideName), csize_t(requiredSize + 1)) != high(csize_t):
           var ptrToEnv = c_getenv(cast[cstring](buf2))
           ptrToEnv[0] = '\0'
           ptrToEnv = c_getenv(cast[cstring](buf2))

--- a/tests/misc/t21443.nim
+++ b/tests/misc/t21443.nim
@@ -1,0 +1,6 @@
+import std/envvars
+
+# bug #19292
+putEnv("NimPutEnvTest", "test")
+# bug #21122
+doAssert getEnv("NimPutEnvTest") == "test"

